### DISCUSSION TITLE
Implementing a Complete Closed-Loop Control Process for Deformable Objects in SOFA

### DISCUSSION THREAD

**Discussion by sodasoda520** [sofa-framework/sofa#5095] [BeamAdapter]
Hello SOFA Community,

I am looking to implement a complete automatic control process for a deformable object within the SOFA framework. The objective is to start from an initial state and apply specified displacements to certain nodes at each step, causing deformation.

Here are the specifics of what I am trying to achieve:

Dynamic Displacement Application: At each step of the simulation, I want to apply different amounts of displacement to specific nodes. The displacement values will be calculated based on the current positions of each node.

Closed-Loop Control Simulation: This process should function as a closed-loop control system, meaning that the input displacements at each step depend on the current shape of the object.

Python Integration: I am looking to implement the entire control process in Python, leveraging the capabilities of SOFA.

I have a couple of questions regarding this:

How can I access and retrieve the current shape or position data of the nodes during the simulation?
Are there any existing examples or cases within the SOFA community that illustrate a similar control process? I would love to learn from them.
Any guidance or suggestions would be greatly appreciated!

Thank you in advance for your help!

**Comment by bakpaul** [comment]
Hello. 

In SOFA, objects positions are stored in an specific object : the MechanicalObject. It contains the positions, velocity, forces and so on at each point of your model.  Those are stored in specific data field of the object easily accessible. 

For example doing exactly what you are asking, I cannot think of one, but there are plenty of examples in the repository : https://github.com/sofa-framework/sofa/tree/master/examples 
You can also look at the examples given in those two repository : https://github.com/sofa-framework/Registration (plugin used to impose the position of the model, given external information) and https://github.com/SofaDefrost/SoftRobots (plugin containing components for soft robot simulation and control)

**Reply by sodasoda520** [reply]
Thank you very much!

**Comment by damienmarchal** [comment]
I'm sure it is possible to have closed loop control in Sofa because several of my colleagues at DEFROST are doing such thing.

I think the example I provided in discussion https://github.com/sofa-framework/sofa/discussions/5143
provides a lot of element to get and set value from a mechanical object at each time frame of a simulation.

**Reply by sodasoda520** [reply]
Thank you for your reply
I will study it carefully.
